src/HooglePlus/FilterTest.hs

```diff
@@ -0,0 +1,194 @@
+      map (\x -> (argName x, name)) var
+    extractQualified (ParenA _ qual) = extractQualified qual
+    extractQualified other = throw $ NotSupportedException ("Not able to extract " ++ show other)
+    buildEnv typeEnv (CxSingle _ item) = typeEnv ++ extractQualified item
+    buildEnv typeEnv (CxTuple _ list) = foldr ((++) . extractQualified) typeEnv list
+
+generateRandomValue :: String -> IO String
+generateRandomValue typeName = do
+  result <- runInterpreter $ do {
+    setImports ["Prelude", "Test.QuickCheck"];
+
+    act <- interpret ("do {\
+    \it <- generate (arbitrary :: Gen (" ++ typeName ++ "));\
+    \return $ show it\
+    \}") (as :: IO String);
+
+    liftIO act
+  }
+  case result of
+    Left err -> putStrLn (displayException err) >> return "error \"114514\""
+    Right res -> return ("(" ++ res ++ ")")
+
+mapRandomMultiple :: (TypeEnv, [ArgumentType], ArgumentType) -> Int -> IO [String]
+mapRandomMultiple env count = do
+  base <- mapRandomArguments' env
+  replicateM count (derive base)
+  where
+    derive :: [(String, String)] -> IO String
+    derive base = do
+      rand <- generate (arbitrary :: Gen Int)
+      let index = rand `rem` length base
+      let item = getNthElement index base
+
+      replaced <- newRandomValue item
+      let list = setNthElement index replaced base
+      return $ unwords (map snd list)
+
+    newRandomValue (name, _) = do
+      val <- generateRandomValue name
+      return (name, val)
+
+    getNthElement idx list = item
+      where (_, item:_) = splitAt idx list
+
+    setNthElement idx item list
+      | idx < 0   = list
+      | otherwise = case splitAt idx list of
+                    (front, _:back) -> front ++ item : back
+                    _ -> list
+
+mapRandomArguments :: (TypeEnv, [ArgumentType], ArgumentType) -> IO String
+mapRandomArguments env = do
+  result <- mapRandomArguments' env
+  return $ unwords (map snd result)
+
+mapRandomArguments' :: (TypeEnv, [ArgumentType], ArgumentType) -> IO [(String, String)]
+mapRandomArguments' (typeEnv, args, retType) = mapM f (transformType typeEnv args)
+  where
+    f x = let str = show x in do
+      val <- generateRandomValue str
+      return (str, val)
+
+    transformType [] args = args
+    transformType (envItem:env) args = transformType env args'
+      where
+        args' = map (replace envItem) args
+
+        replace (Polymorphic name, cons) (Polymorphic name') =
+          if name == name' then
+            case cons of
+              "Eq" -> Concrete "Int"
+              "Num" -> Concrete "Int"
+          else Polymorphic name'
+        replace cons (ArgTypeList argType) = ArgTypeList (replace cons argType)
+        replace _ x = x
+
+eval_ :: [String] -> String -> Int -> IO (Maybe (Either InterpreterError String))
+eval_ modules line time = timeout time $ runInterpreter $ do {
+  setImports modules;
+  eval line
+}
+
+runChecks :: Environment -> RType -> UProgram -> IO Bool
+runChecks env goalType prog = do
```

Comment:
I'd like to have it as some top-most check function (or entry point) for the test-based filter module. Perhaps we could separate `progDoesNotCrash` as a new function, or there's any better suggestions?

Reply:
Sure, it can still be top level and exported. We're not implementing an interface here where we need to maintain a name--it can change without much trouble.